### PR TITLE
util: new XIVAPI CLI query tool

### PR DESCRIPTION
**Pull request by wexxlee**
Adds a new XIVAPI CLI query tool to util.  It's a lightweight script that fetches endpoint data, filters it based on a user-specified filter string, and outputs it as JSON to the console.  I put an early version of this together for looking into some of the data issues during the util script cleanup, and it was pretty helpful.  It's grown a bit since. 👀

More info on usage available by running `npm run query -- -h`.

Since it uses regex on user-supplied inputs, there's a lot of error handling needed to capture wonky user input.  I *think* I've accounted for most of the possibilities, but open to anything I missed.

Comments/suggestions welcome!

### CODE REVIEW

File: util/query_xivapi.ts
Comment (valarnin) [inline util/query_xivapi.ts:180]:
```suggestion
    FILTER:   If the --filter option is used, it should be in the format [column] [operator] [value].
```

File: util/query_xivapi.ts
Comment (valarnin) [inline util/query_xivapi.ts:283]: There's a lot of effort here to parse out filter information. Would it make more sense to just have `--filters` accept a JSON object, for example something like these:
```json
[{"column":"ID","operator":">","value":30}]
[{"column":"ID","operator":">","value":30},{"column":"Name","operator":"~","value":["Topaz Titan","Ruby Carb*"]}]
{"column":"ID","operator":">","value":30}
```

(These could also be "JSON-like", as was done for timeline netregex parsing)

Mostly I suggest this because regex is a pain to debug later down the line whenever it gets even slightly complicated.

File: util/xivapi.ts
Comment (valarnin) [inline util/xivapi.ts:100]: The error message here changed, was this intentional? And if so, and this is a more generic error, is there any additional information we can glean from `response` to also bubble up on the thrown `Error`?

Comment (wexxlee) [reply util/xivapi.ts:100]: I dug into this more when cleaning up the CLI tool, and `response.ok` is solely based on whether the HTTP status code returned from the fetch is 200-299.  So calling it a "network" error seemed potentially inaccurate, since it could be other things.  The main change here was moving `jsonResult = (await response.json()) as XivApiResult;` above the error throw so that we actually capture the JSON output if there is an error.  Now, when you generate an XIVAPI error (say by hitting a non-existent endpoint or node), the console will actually show real info, e.g.:
```json
{
  "Error": true,
  "Subject": "XIVAPI ERROR",
  "Note": "Get on discord: https://discord.gg/MFFVHWC",
  "Message": "Game Data does not exist: Pet 85",
  "Hash": "4fb92446196ac6bba6f22a3b6964c5c20ae7ef23",
  "Ex": "Symfony\\Component\\HttpKernel\\Exception\\NotFoundHttpException",
  "ExCode": 404,
  "Url": "",
  "Debug": {
    "ID": "1UTBAx1E24010710",
    "File": "#53 xivapi.com/src/Service/Content/GameData.php",
    "Method": "GET",
    "Path": "/Pet/85",
    "Action": "App\\Controller\\XivGameContentController::contentData",
    "Code": 404,
    "Date": "2024-01-07 22:27:57",
    "Env": "prod"
  }
}
```

File: util/query_xivapi.ts
Comment (wexxlee) [reply util/query_xivapi.ts:283]: > There's a lot of effort here to parse out filter information. Would it make more sense to just have `--filters` accept a JSON object, for example something like these:

I've noodled on this for a bit, but I think I lean toward keeping as-is.  Changing to JSON-style input would remove the initial capture regex (although some will still be needed to handle wildcards).  But most of the code in that `parseFilters` will still be needed to validate user input, regardless of whether it comes as a string or JSON.  If anything, the regex saves some work by rejecting most non-conforming input.

I guess I see higher value in the simplicity of short text for quick CLI input, rather than typing out (what could be rather lengthy) JSON.  If there were a text editor involved, I think it'd be easier, but a CLI makes a lengthy string hard to read, check for errors, and edit after a certain length.  Particularly if there's multiple filters, that's a lot of characters devoted to repeating 'column', 'operator', 'value' over and over.

Comment (JLGarber) [reply util/query_xivapi.ts:283]: I think I lean mildly toward "keep it as it is and do not change it to accept JSON-likes" myself. I agree with valarnin that maintenance on regex-based parsing can be a real pain, but as a user of the utility, it's a lot easier to use if I don't have to worry about formatting a bunch of JSON on the command line. Particularly since we're looking at 7.0 in 6-8 months, I think it's worth taking on a little more burden for maintainers in order to save a little time later when *using* the script in the new-content time crunch.

Comment (valarnin) [reply util/query_xivapi.ts:283]: Yeah, I've given this some more thought over the past couple days and I think it's fine to keep it as is.

Maybe we should encourage adding comments with specific examples of "match" and "don't match" for any regex above a certain complexity going forward?

**Review by valarnin — APPROVED** [review]
Approving this for now. Feel free to merge at your convenience, depending on thoughts on comments for the regexes above.

File: util/query_xivapi.ts
Comment (wexxlee) [reply util/query_xivapi.ts:283]: > Maybe we should encourage adding comments with specific examples of "match" and "don't match" for any regex above a certain complexity going forward?

Yup, makes sense to me.  I think the filter separator and the value-array regex are both fugly enough to warrant comments, so I'll add some before merging.  I think the others are probably simple enough not to need comments, but maybe an overall comment on the structure of a proper filter would be useful.

**Pull request merged**
